internal/controller/promotions/promotions.go

```diff
@@ -493,11 +499,13 @@ func (r *reconciler) promote(
 	if err != nil {
 		return err
 	}
-	stage.Status.CurrentState = &nextState
-	stage.Status.AvailableStates[targetStateIndex] = nextState
-	stage.Status.History.Push(nextState)
 
-	err = r.kargoClient.Status().Update(ctx, stage)
+	err = kubeclient.PatchStatus(ctx, r.kargoClient, stage, func(status *api.StageStatus) {
+		status.CurrentState = &nextState
+		status.AvailableStates[targetStateIndex] = nextState
```

Comment:
I have the same understanding. We should be safe from race conditions because the controller does not process multiple promotions in one stage.

Added comment about it.